File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
import java.io.*;
import java.util.*;

class Solution {
    public int solution(int[] diffs, int[] times, long limit) {
        long time = limit+1;
        int max = 100000;
        int min = 1;
        
        int level = 0;
        while(max >= min){
            level = (max + min) / 2;
            time = 0;
            for(int i = 0; i<diffs.length; i++){
                int l = diffs[i] - level;
                if(l <= 0){
                    time += times[i];
                }else{
                    time += (times[i-1] + times[i]) *l + times[i];
                }
            }
            
            if (time > limit){
                min = level + 1;
            }else{
                max = level - 1 ;
            }

        }

        int answer = min;
        return answer;
    }
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def solution(diffs, times, limit):

    start = 1
    end  = max(diffs)

    while(start <= end):
        mid = (start + end)//2
        time = 0
        for i in range(len(diffs)):
            if mid < diffs[i]:
                time+= (diffs[i] - mid)*(times[i-1]+times[i]) + times[i]
            else:
                time += times[i]
        if time ==limit:
            start = mid
            break
        elif time <limit:
            end = mid - 1
        else:
            start = mid+1
        
                   
    return start